### ISSUE TITLE
Cast to Open Generic

### ISSUE THREAD

**Issue by iskiselev** [sq/JSIL#785]
Here is test case:
```
using System;

public interface Interface1<T>
{
    
}

public interface Interface2<T>
{

}

public class GenericClass<T> : Interface1<T>, Interface2<T>
{
    public bool Test(Interface1<T> input)
    {
        var asInterface2 = input as Interface2<T>;
        if (asInterface2 != null)
        {
            return this == asInterface2;
        }
        return false;
    }
}

public static class Program {
    public static void Main (string[] args) {
        var instance = new GenericClass<object>();
        Console.WriteLine(instance.Test(instance) ? "True" : "False");
    }
}
```
And here is translation of `Test` function:
```
  function GenericClass$b1_Test (input) {
    var asInterface2 = $asm00.Interface2$b1.Of($thisType.T.get(this)).$As(input);
    return ((asInterface2 !== null) && (this === $thisType.$Cast(asInterface2)));
  };
```

JSIL tries to cast object to `$thisType`, but `$thisType` is open generic. Bad news is that ILSpy set open generic as ExpectedType of comparison operand. Will think how to fix/workaround it, as this pattern is used in BCL `List<T>` implementation.

**Comment by kg** [comment]
Yeah, ILSpy type inference screws up a lot of these cases. I have workarounds for some of them already

**Comment by iskiselev** [comment]
I've found two options of fixing this problem (and will create PR to discuss it), but really we have here also (or maybe only) anther problem, that is same with #770.
In translated sources JSIL tries to downcast Interface2 to GenericClass, which may be impossible.
I'm not sure if we need cast here at all.

**Comment by iskiselev** [comment]
Really much simpler test case:
```
using System;

public interface Interface1<T>
{ 
}

public class GenericClass<T> : Interface1<T>
{
    public bool Test(Interface1<T> input)
    {
            return this == input;
    }
}

public static class Program {
    public static void Main (string[] args) {
        var instance = new GenericClass<object>();
        Console.WriteLine(instance.Test(instance) ? "True" : "False");
    }
}
```

**Comment by iskiselev** [comment]
Fixed via #786.
